Add `workflow.output` description to supported operations guide (#3294)
thin-edge/thin-edge.io: docs/src/operate/c8y/supported-operations.md

The mapper detects the change in the `/etc/tedge/operations/c8y` directory except for `.template` file. When the change is detected, the SmartREST `114` (supported operation) message is sent to Cumulocity.

After that, the received JSON over MQTT message is converted to %%te%% command, preserving all parameters provided in `input` field.

> The payload also contains the parameters retrieved from the operation template file, notably `on_fragment` and `outout`.

Reviewer: Minor typo.

```suggestion
The payload also contains the parameters retrieved from the operation template file, notably `on_fragment` and `output`.
```

> The payload also contains the parameters retrieved from the operation template file, notably `on_fragment` and `outout`.

Author: Already fixed in this thread https://github.com/thin-edge/thin-edge.io/pull/3294#discussion_r1883559220

```json5 title="Payload"
{
  "status": "init",
  "text":"echo helloworld",
  "foo":"bar"
}
```

Make sure that provided input is JSON object, otherwise the operation execution will be skipped.  

:::

:::info

> The `output` field is required for specific operations (e.g., `c8y_Command` and `c8y_RelayArray`) that need to parse the operation result upon successful execution. This field accepts either a `String` or an `Array` value. Placeholders in the output will be replaced with the actual values and appended as parameters to the SmartREST `503`/`506` messages.

Reviewer 2: ```suggestion
The `output` field is required for specific operations (e.g., `c8y_Command` and `c8y_RelayArray`) that needs to send additional parameters to the cloud to complete an operation successfully.
These parameters are extracted from the operation payload using the values specified in the `output`field.
This field accepts either a `String` or an `Array` value.
Placeholders in the output will be replaced with the actual payload values and appended as parameters to the SmartREST `503`/`506` messages.
```

#### Example 1: c8y_Command

For the `c8y_Command` operation, the execution output is stored in the `result` field. Here is an example JSON payload when it is successful:
```json5 title="c8y_Command"
{
  "c8y-mapper": {
    "on_fragment": "c8y_Command",
    "output": "${.payload.result}"
  },
  "command":"echo helloworld",
  "result":"helloworld",
  "status": "successful"
}
```

This message is converted to the following SmartREST message:
```
503,c8y_Command,helloworld
```

#### Example 2: c8y_RelayArray

For the c8y_RelayArray operation, the output is an array. Here is an example JSON payload when it is successful:

```json5 title="c8y_RelayArray"
{
  "c8y-mapper": {
    "on_fragment": "c8y_RelayArray",
    "output": "${.payload.result}"
  },
  "command":"echo helloworld",

>   "result":["OPEN","CLOSED"],

Reviewer 2: ```suggestion
  "states":["OPEN","CLOSED"],
```
May be we can use a different key in this second example to convey the point that it doesn't always have to be named `result`.

Available only in the template operation file definition:
* `workflow.operation` - The command name that will trigger workflow execution.
* `workflow.input` - The JSON object input that can be used in the workflow.

> * `workflow.output` - The JSON object that is converted to the additinal parameters of the SmartREST successful status update messages (`501`/`504`)

Reviewer: Fixing minor type (not picked up by the typo checker unfortunately)

```suggestion
* `workflow.output` - The JSON object that is converted to the additional parameters of the SmartREST successful status update messages (`501`/`504`)
```

Though are you sure the description is correct, doesn't the `workflow.output` get included in the 503/506 messages (e.g. setting the operation to `SUCCESSFUL` status and setting the result).

Author: True, 503/506 messages :D

Author: fixed in [14aa737](https://github.com/thin-edge/thin-edge.io/pull/3294/commits/14aa737582ad996088c50bd582dd50e372c7f811)
